// stylelint-codeguide/dist/rules/selector-max-empty-lines/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ruleName = exports.meta = exports.messages = exports["default"] = void 0;
var _report = _interopRequireDefault(require("../../utils/report"));
var _ruleMessages = _interopRequireDefault(require("../../utils/ruleMessages"));
var _validateOptions = _interopRequireDefault(require("../../utils/validateOptions"));
var _validateTypes = require("../../utils/validateTypes");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }
var ruleName = "selector-max-empty-lines";
exports.ruleName = ruleName;
var messages = (0, _ruleMessages["default"])(ruleName, {
  expected: function expected(max) {
    return "Expected no more than ".concat(max, " empty ").concat(max === 1 ? "line" : "lines");
  }
});
exports.messages = messages;
var meta = {
  url: "https://github.com/firefoxic/stylelint-codeguide/blob/main/lib/rules/selector-max-empty-lines/README.md",
  fixable: true
};

/** @type {import('stylelint').Rule} */
exports.meta = meta;
var rule = function rule(primary, _secondaryOptions, context) {
  var maxAdjacentNewlines = primary + 1;
  return function (root, result) {
    var validOptions = (0, _validateOptions["default"])(result, ruleName, {
      actual: primary,
      possible: _validateTypes.isNumber
    });
    if (!validOptions) {
      return;
    }
    var violatedCRLFNewLinesRegex = new RegExp("(?:\r\n){".concat(maxAdjacentNewlines + 1, ",}"));
    var violatedLFNewLinesRegex = new RegExp("\n{".concat(maxAdjacentNewlines + 1, ",}"));
    var allowedLFNewLinesString = context.fix ? "\n".repeat(maxAdjacentNewlines) : "";
    var allowedCRLFNewLinesString = context.fix ? "\r\n".repeat(maxAdjacentNewlines) : "";
    root.walkRules(function (ruleNode) {
      var selector = ruleNode.raws.selector ? ruleNode.raws.selector.raw : ruleNode.selector;
      if (context.fix) {
        var newSelectorString = selector.replace(new RegExp(violatedLFNewLinesRegex, "gm"), allowedLFNewLinesString).replace(new RegExp(violatedCRLFNewLinesRegex, "gm"), allowedCRLFNewLinesString);
        if (ruleNode.raws.selector) {
          ruleNode.raws.selector.raw = newSelectorString;
        } else {
          ruleNode.selector = newSelectorString;
        }
      } else if (violatedLFNewLinesRegex.test(selector) || violatedCRLFNewLinesRegex.test(selector)) {
        (0, _report["default"])({
          message: messages.expected(primary),
          node: ruleNode,
          index: 0,
          result: result,
          ruleName: ruleName
        });
      }
    });
  };
};
rule.ruleName = ruleName;
rule.messages = messages;
rule.meta = meta;
var _default = rule;
exports["default"] = _default;
